esp32 i2s driver improvement and i2s_api test enabling

wmrsouza opened the pull request:
This PR promotes the following improvements to the esp32 i2s driver:
- fix issue regarding last block received being lost
- make additional checks during stream initialization
- fix reduntant pin configuration
- allow deferred processing of empty tx queue
- make consistent use of the driver state inside i2s_read and i2s_write apis
- add missing configuration checks
- change log error messages to log debug messages to avoid timing issues
- optimize stream structs
- add sanity checks
- correct driver state management
- correct I2S_DIR_BOTH handling
- enable i2s_api test for: esp32c3_devkitm, esp32c6_devkitc_hpcore, esp32h2_devkitm, and esp32s3_devkitm_procpu
- implement simplified i2s test for: esp32_devkitc_procpu, and esp32s2_devkitc

wmrsouza committed:
tests: drivers: i2s_api: add esp32 boards

adds the following boards ot i2s_api test:
- esp32c3_devkitm
- esp32c6_devkitc/esp32c6/hpcore
- esp32s3_devkitc/esp32s3/procpu
- esp32h2_devkitm

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

wmrsouza committed:
drivers: i2s: esp32: fix last block loss

Avoids erroneous memory block freeing when stopping transfers

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

wmrsouza committed:
drivers: i2s: consistent use of stream state

Reworks i2s_esp32_read() and i2s_esp32_write() to allow consistent use of
stream->data->state

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

wmrsouza committed:
drivers: i2s: esp32: add missing checks

Adds missing checks regarding mem_slab and data_format during each stream
configuration

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

wmrsouza committed:
drivers: i2s: esp32: log messages rework

Reworks log messages to avoid timing issues

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

wmrsouza committed:
drivers: i2s: esp32: fix redundant pin configuration

Moves the pinctrl_apply_state() call to avoid calling it twice

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

wmrsouza committed:
drivers: i2s: esp32: allow deferred processing of an empty tx queue

Allows to defer processing of an empty tx queue when tx_callback is
executed. This prevents error conditions if the tx dma interrupt is
generated too early, as occurs with esp32 and esp32s2.

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

wmrsouza committed:
tests: boards: esp32: add simplified i2s test

Adds simplified i2s test for the following platforms:
- esp32_devkitc/esp32/procpu
- esp32s2_saola

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

sylvioalves commented on drivers/i2s/i2s_esp32.c:
```diff
@@ -108,12 +112,12 @@ static esp_err_t i2s_esp32_calculate_clock(const struct i2s_config *i2s_cfg, uin
 
 	if (i2s_cfg == NULL) {
-		LOG_ERR("Input i2s_cfg is NULL");
+		LOG_DBG("Input i2s_cfg is NULL");
```
I think you can remove all those logs related to NULL pointers. It already returns INVALID_ARGS.

sylvioalves commented on drivers/i2s/Kconfig.esp32:
```diff
@@ -31,4 +31,10 @@ config I2S_ESP32_DMA_DESC_NUM_MAX
 	  Max number of link descriptor available for DMA transfers on each I2S channel
 
+config I2S_ESP32_ALLOWED_EMPTY_TX_QUEUE_DEFERRAL_TIME_MS
+	int "ESP32 I2S number of link descriptors"
```
`int "ESP32 I2S number of link descriptors"`
The description does not fit the config itself, which is time related.

sylvioalves replied:
is 0 a good default value? It means `k_timer_start` will start it without any waiting interval.

sylvioalves commented on tests/boards/espressif/i2s/src/main.c:
```diff
@@ -0,0 +1,248 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Why do we need this customized test case when `i2s_additional` test already allows gpio_looback scenario?

wmrsouza replied:
The existing I2S Zephyr tests assume a synchronous start for TX and RX, which isn't perfectly aligned on ESP32 and ESP32-S2 socs due to a hardware constraint. This issue creates a lag on the first block received that propagates to all following blocks making the existing I2S test to fail.

This platform-specific test modifies the flow to accommodate the behavior described.  It verifies successful data transfer for the blocks that are received (n-1 blocks) without attempting to read the nth block that was sent last, thus providing a reliable pass/fail indicator for I2S loopback on these targets.

wmrsouza committed:
drivers: i2s: esp32: optimize stream structs and add sanity check

Optimizes i2s_esp32_stream_data and i2s_esp32_stream_conf structs and adds
property sanity check according to soc series based on SOC_GDMA_SUPPORTED
feature

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

raffarost reviewed (dismissed):
good job

wmrsouza committed:
boards: esp32: rework i2s pin config on dtsi, dts, and overlay files

Reworks i2s entries on esp32 pinctrl.dtsi and .dts board files and
adequates overlay files regarding to i2s samples

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

thedjnK reviewed (dismissed):
LED strip update OK

wmrsouza committed:
drivers: i2s: esp32: fix driver state management and I2S_DIR_BOTH behavior

Rework to fix state management and I2S_DIR_BOTH behavior across the driver

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

sylvioalves reviewed (changes requested):
Please, add ESP32 platforms allowed in this test case like:
```c
    platform_allow:
      - esp32_devkitc/esp32/procpu
      - esp32c3_devkitm
      - esp32c6_devkitc/esp32c6/hpcore
      - esp32h2_devkitm
      - esp32s2_saola
      (...)
```

wmrsouza commented:
> Please, add ESP32 platforms allowed in this test case like:
> 
> ```c
>     platform_allow:
>       - esp32_devkitc/esp32/procpu
>       - esp32c3_devkitm
>       - esp32c6_devkitc/esp32c6/hpcore
>       - esp32h2_devkitm
>       - esp32s2_saola
>       (...)
> ```

I can do this for `zephyr/tests/boads/espressif/i2s`, but it should not be done for `zephyr/tests/drivers/i2s/i2s_api` because in this case boards will be selected for `drivers.i2s.gpio_loopback` based on `filter: CONFIG_I2S_TEST_USE_GPIO_LOOPBACK`

thedjnK reviewed (approved):
LED strip updates OK
